### PR TITLE
Support FreeBSD

### PR DESCRIPTION
**Pull request by turtleli**
It more or less works on FreeBSD 10.3, I only did light testing though. I compiled with clang 3.7, didn't test other compilers.

I couldn't get my DS3 to work. I've only had FreeBSD installed on my machine for 2 days though, so I don't know whether it's SDL2 joystick support/FreeBSD/PEBKAC and whether there are workarounds.

The PR needs some rethinking/improvements (there should definitely be a better way of doing the cmake stuff and ifdefs) plus more testing.

### CODE REVIEW

**Comment by turtleli** [comment]
Updated a bit.

Main changes from initial version:
 * CMake - Use UNIX (note: true for OS X) most of the time
 * Used \__unix__ for ifdefs - Unlike CMake, it doesn't include OS X (I could be wrong)
 * Used alignas for gsdx __aligned
 * Added intrinsics yuv2rgb implementation (will be used for non-Linux clang users) - variable naming sucks at the moment, but it does work. Converted from the optimised assembly version.

Notes (partly for my own benefit cause I'll forget):
 * aio kernel module needs to be loaded.
 * aio_error busy loop - sleep it?
 * sVU memory reservation always fails (both VU0 and VU1).
 * ALSA needs to be hidden from SPU2-X GUI
 * Some \__linux__ defines haven't been checked yet
 * Remove hardcoded clang version from build.sh
 * GCC
 * Check from Linux/Windows to see if anything can be made simpler
 * Rename files to reflect purpose after PR?

**Comment by gregory38** [comment]
great work. :+1: 

```
it doesn't include OS X (I could be wrong)
```
Hum, try to do a ` grep _unix_  -r | grep define` in pcsx2 or common. Oh wait maybe they used _posix_ for mac stuff. I don't remember.

```
sVU memory reservation always fails (both VU0 and VU1).
```
Honestly better remove it (at least for a first time)

**Comment by turtleli** [comment]
It seems \__POSIX__ is used, though we define that ourselves in common/include/Pcsx2Defs.h:
```
#if defined(__linux__) || defined(__APPLE__) || defined(__unix__) || defined(__CYGWIN__) || defined(__LINUX__)
#	if !defined(__POSIX__)
#		define __POSIX__ 1
#	endif
#endif
```

Yeah, I probably will just disable sVU.

**Comment by MrCK1** [comment]
Sorry to do this (figured you might be online as I write this)

All PCSX2 related servers are not responding (wiki, forums, pcsx2.net) didn't know if anybody had a handle on this yet

**Comment by turtleli** [comment]
On the yuv2rgb implementations - I think I might remove the gcc inline assembly version and just use the intrinsics version - I checked sse2, sse4.1, avx, and avx2 builds on Linux, and the performance seemed the same (the intrinsics version felt ever so slightly faster (~0.5fps), though I could be imagining things). I need to check Windows performance.

**Comment by gregory38** [comment]
In this case, I suggest to use posix for parts that are common between linux/freebsb/mac.

yuv2rgb, yeah perf ought to be the same. Hum to be sure, you could also try the cpp version if there is any speed impact.

**Comment by turtleli** [comment]
> yuv2rgb, yeah perf ought to be the same. Hum to be sure, you could also try the cpp version if there is any speed impact.

Do you mean the reference version? fps drops from ~120 to ~73 on my tests.

**Comment by gregory38** [comment]
Yes. It was to know the speed impact of the code. I'm curious if an AVX port might help to squeeze some extra fps. Anyway better keeps the code clean. It is a good jobs.

After the merge, it would be interesting to check what are the remaining asm code. Maybe we could remove the `-no-integrated-as` option from cmake.

**Comment by turtleli** [comment]
Oh, I actually tried an AVX2 port - I don't know whether it was fully optimised, but I managed an extra 2-3fps.

I checked Windows, like Linux it seems marginally better. I'm guessing its due to instruction reordering. So yeah, I'll be removing the Windows assembly version too.

`-no-integrated-as` probably could be removed - I had to remove the flag for FreeBSD clang 3.7 since the external assembler was too old.

**Comment by gregory38** [comment]
Oh ok. Maybe add a note of the AVX2 impact.

**Comment by turtleli** [comment]
Changes since last update:
 * SPU2-X - SDL backend temporarily disabled on FreeBSD - it'll make the SPU2-X config dialog code extremely messy with ifdefs.
 * IOP gte - used count_leading_sign_bits to replace some assembly
 * yuv2rgb sse2 intrinsics implementation more readable, assembly versions removed.
 * gsdx: removed use of ftime - man pages + POSIX specs all suggest it's obsolete. clock_gettime is used instead.
 * Disabled SuperVU in cmake for all non Linux OSes
 * Improved POSIX AIO iso reading implementation
 * Removed the build.sh commit (commit was so I could be lazy)

The main thing which is missing is probably joypad support (maybe it's just my setup), but the PR is probably good enough for now.

File: common/src/Utilities/Linux/LnxThreads.cpp [hunk 1]
```diff
@@ -113,9 +117,13 @@ void Threading::pxThread::_platform_specific_OnCleanupInThread()
 
 void Threading::pxThread::_DoSetThreadName( const char* name )
 {
+#if defined(__linux__)
 	// Extract of manpage: "The name can be up to 16 bytes long, and should be
 	//						null-terminated if it contains fewer bytes."
 	prctl(PR_SET_NAME, name, 0, 0, 0);
+#elif defined(__unix__)
+	pthread_set_name_np(pthread_self(), name);
```
Comment (gregory38) [inline common/src/Utilities/Linux/LnxThreads.cpp]: Could `pthread_set_name_np` work on linux too ?

**Comment by gregory38** [comment]
Except for the small comment. I'm fine with the PR. Better merge it now before I broke others stuffs.

File: common/src/Utilities/Linux/LnxThreads.cpp (hunk 1 above)
Comment (turtleli) [inline common/src/Utilities/Linux/LnxThreads.cpp]: `pthread_set_name_np` - FreeBSD/OpenBSD
`pthread_setname_np` - Linux/NetBSD

Great fun ;) But yeah could be worth changing later.

**Pull request merged**